### PR TITLE
codegen.ml: generating printf function and function_decls

### PR DESCRIPTION
**Pull request by andylhxu**

### CODE REVIEW

**Comment by andylhxu** [comment]
build_function_body is still incompleted. Main function termination token not generated correctly.

File: src/frontend/ast.mli [hunk 1]
```diff
@@ -8,6 +8,7 @@ type typ =
   | Struct of string (* struct identifier *)
   | Array of typ
   | Vector of int (* number of elements in vector *)
+  | Void (* ANDYTODO: review this *)
```
Comment (Contentmaudlin) [inline src/frontend/ast.mli]: Wait why did we add a Void type?

Comment (andylhxu) [reply src/frontend/ast.mli]: we have to tell llvm function generator the return type of all functions. Can't be optional. I'll think about ways to resolve this now.

**Review by Contentmaudlin — COMMENTED** [review]
Crashes on hello world?

**Review by Contentmaudlin — APPROVED** [review]
LGTM for now

**Comment by andylhxu** [comment]
yep, let's not merge it yet for now.

File: src/frontend/ast.mli (hunk 1 above)
Comment (Contentmaudlin) [reply src/frontend/ast.mli]: Can't you check for `None` and match it to a LLVM Void type? So only have a Void in the backend?

**Comment by andylhxu** [comment]
Not using Blocks now.

File: src/frontend/ast.mli [hunk 2]
```diff
@@ -59,7 +59,7 @@ and expr =
   | Cond of expr * expr * expr (* technically Ternop *)
 
 and stmt =
-  | VDecl of bind * expr option
+    VDecl of bind * expr option
```
Comment (Contentmaudlin) [inline src/frontend/ast.mli:62]: Fucking put back the `|`

**Review by Contentmaudlin — CHANGES_REQUESTED** [review]
Please put back the `|` in `parser.mly`

**Review by Contentmaudlin — APPROVED** [review]

**Pull request merged**